substrate/frame/delegated-staking/src/types.rs

```diff
@@ -118,10 +128,24 @@ impl<T: Config> AgentLedger<T> {
 	}
 
 	/// Save self to storage with the given key.
+	///
+	/// Increments provider count if this is a new agent.
 	pub(crate) fn update(self, key: &T::AccountId) {
+		if !<Agents<T>>::contains_key(key) {
+			// This is a new agent. Provide for this account.
+			frame_system::Pallet::<T>::inc_providers(key);
+		}
 		<Agents<T>>::insert(key, self)
 	}
 
+	/// Remove self from storage.
+	pub(crate) fn remove(key: &T::AccountId) {
+		debug_assert!(<Agents<T>>::contains_key(key), "Agent should exist in storage");
+		<Agents<T>>::remove(key);
```

Comment:
```suggestion
		Agents::<T>::remove(key);
```

(same suggestion for `fn update`)